pkg/api/defaults.go

```diff
@@ -508,6 +512,69 @@ func (p *Properties) setHostedMasterProfileDefaults() {
+			EndpointConfig: AzureEndpointConfig{
+				ResourceManagerVMDNSSuffix: "",
+			},
+			OSImageConfig: map[Distro]AzureOSImageConfig{
+				Ubuntu:          DefaultUbuntuImageConfig,
+				RHEL:            DefaultRHELOSImageConfig,
+				CoreOS:          DefaultCoreOSImageConfig,
+				AKS:             DefaultAKSOSImageConfig,
+				AKSDockerEngine: DefaultAKSDockerEngineOSImageConfig,
+			},
+		}
+
+		//Set default value for ResourceManagerVMDNSSuffix
+		azureStackCloudSpec.EndpointConfig.ResourceManagerVMDNSSuffix = p.CustomCloudProfile.Environment.ResourceManagerVMDNSSuffix
+
+		// Use the custom input to overwrite the default values in AzureStackCloudSpec
+		if p.CustomCloudProfile.AzureEnvironmentSpecConfig != nil {
+			helpers.AssignStringIfNotEmpty(&p.CustomCloudProfile.AzureEnvironmentSpecConfig.CloudName, p.CustomCloudProfile.AzureEnvironmentSpecConfig.CloudName)
+
+			// DockerSpecConfig
+			helpers.AssignStringIfNotEmpty(&azureStackCloudSpec.DockerSpecConfig.DockerComposeDownloadURL, p.CustomCloudProfile.AzureEnvironmentSpecConfig.DockerSpecConfig.DockerComposeDownloadURL)
+			helpers.AssignStringIfNotEmpty(&azureStackCloudSpec.DockerSpecConfig.DockerEngineRepo, p.CustomCloudProfile.AzureEnvironmentSpecConfig.DockerSpecConfig.DockerEngineRepo)
+
+			//KubernetesSpecConfig
+			helpers.AssignStringIfNotEmpty(&azureStackCloudSpec.KubernetesSpecConfig.ACIConnectorImageBase, p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.ACIConnectorImageBase)
+			helpers.AssignStringIfNotEmpty(&azureStackCloudSpec.KubernetesSpecConfig.AzureCNIImageBase, p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.AzureCNIImageBase)
+			helpers.AssignStringIfNotEmpty(&azureStackCloudSpec.KubernetesSpecConfig.CNIPluginsDownloadURL, p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.CNIPluginsDownloadURL)
+			helpers.AssignStringIfNotEmpty(&azureStackCloudSpec.KubernetesSpecConfig.ContainerdDownloadURLBase, p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.ContainerdDownloadURLBase)
+			helpers.AssignStringIfNotEmpty(&azureStackCloudSpec.KubernetesSpecConfig.EtcdDownloadURLBase, p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.EtcdDownloadURLBase)
+			helpers.AssignStringIfNotEmpty(&azureStackCloudSpec.KubernetesSpecConfig.KubeBinariesSASURLBase, p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.KubeBinariesSASURLBase)
+			helpers.AssignStringIfNotEmpty(&azureStackCloudSpec.KubernetesSpecConfig.KubernetesImageBase, p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.KubernetesImageBase)
+			helpers.AssignStringIfNotEmpty(&azureStackCloudSpec.KubernetesSpecConfig.NVIDIAImageBase, p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.NVIDIAImageBase)
+			helpers.AssignStringIfNotEmpty(&azureStackCloudSpec.KubernetesSpecConfig.TillerImageBase, p.CustomCloudProfile.AzureEnvironmentSpecConfig.KubernetesSpecConfig.TillerImageBase)
```

Comment:
Nice-to-have: it would be nice to have a separate private method for each struct inside of `azureStackCloudSpec`.

i.e. - a separate private method for populating `KubernetesSpecConfig` defaults, `DockerSpecConfig` defaults etc.